src/Middleware/RequestAuthorizationMiddleware.php

```diff
@@ -0,0 +1,178 @@
+
+/**
+ * Request Authorization Middleware
+ *
+ * This MUST be added after the Authorization, Authentication and
+ * RoutingMiddleware in the Middleware Queue!
+ *
+ * This middleware is useful when you want to authorize your requests, for example
+ * each controller and action, against a role based access system or any other
+ * kind of authorization process that controls access to certain actions.
+ */
+class RequestAuthorizationMiddleware
+{
+
+    use InstanceConfigTrait;
+
+    /**
+     * Default Config
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'authorizationAttribute' => 'authorization',
+        'identityAttribute' => 'identity',
+        'redirectUrl' => null,
+        'unauthorizedHandler' => null,
+        'missingIdentityHandler' => null,
+        'allowAccessWhenIdentityIsMissing' => false,
+        'canMethod' => 'access'
+    ];
+
+    /**
+     * Constructor
+     *
+     * @param array $config Configuration options
+     */
+    public function __construct($config = [])
+    {
+        $this->setConfig($config);
+    }
+
+    /**
+     * Gets the authorization service from the request attribute
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request Server request.
+     * @return \Authorization\AuthorizationServiceInterface
+     */
+    protected function getServiceFromRequest(ServerRequestInterface $request)
+    {
+        $serviceAttribute = $this->getConfig('authorizationAttribute');
+        $service = ($request->getAttribute($serviceAttribute));
+
+        if (!$service instanceof AuthorizationServiceInterface) {
+            $errorMessage = __CLASS__ . ' could not find the authorization service in the request attribute. '
+                . 'Make sure you added the AuthorizationMiddleware before this middleware or that you '
+                . 'somehow else added the service to the requests `' . $serviceAttribute . '` attribute.';
+
+            throw new RuntimeException($errorMessage);
+        }
+
+        return $service;
+    }
+
+    /**
+     * Callable implementation for the middleware stack.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request Server request.
+     * @param \Psr\Http\Message\ResponseInterface $response Response.
+     * @param callable $next The next middleware to call.
+     * @return ResponseInterface A response.
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next)
+    {
+        $service = $this->getServiceFromRequest($request);
+        $identity = $request->getAttribute($this->getConfig('identityAttribute'));
+
+        if ($identity instanceof IdentityInterface) {
+            if (!$service->can($identity, $this->getConfig('canMethod'), $request)) {
+                $response = $this->handleUnauthorized($identity, $request, $response);
+            }
+        } else {
+            $response = $this->handleMissingIdentity($request, $response);
```

Comment:
As above - throw `Authorization\Exception\MissingIdentityException` here and let the handler do the rest.